vscode/src/testExplorer.ts

```diff
@@ -0,0 +1,204 @@
+    }
+
+    const programConfig = program.programConfig;
+    const worker = getCommonCompilerWorker(context);
+    const allTestCallables = await worker.getTestCallables(programConfig);
+
+    // only update test callables from this Uri
+    const scopedTestCallables =
+      uri === null
+        ? allTestCallables
+        : allTestCallables.filter(({ location }) => {
+            const vscLocation = toVsCodeLocation(location);
+            return vscLocation.uri.toString() === uri.toString();
+          });
+
+    // break down the test callable into its parts, so we can construct
+    // the namespace hierarchy in the test explorer
+    for (const { callableName, location } of scopedTestCallables) {
+      const vscLocation = toVsCodeLocation(location);
+      const parts = callableName.split(".");
+
+      // for an individual test case, e.g. foo.bar.baz, create a hierarchy of items
+      let rover = ctrl.items;
+      for (let i = 0; i < parts.length; i++) {
+        const part = parts[i];
+        const id = i === parts.length - 1 ? callableName : part;
+        if (!rover.get(part)) {
+          const testItem = ctrl.createTestItem(id, part, vscLocation.uri);
+          testItem.range = vscLocation.range;
+          rover.add(testItem);
+        }
+        rover = rover.get(id)!.children;
+      }
+    }
+  };
+}
+
+/**
+ * Initializes the test explorer with the Q# tests in the active document.
+ **/
+export async function initTestExplorer(
+  context: vscode.ExtensionContext,
+  updateDocumentEvent: vscode.Event<vscode.Uri>,
+) {
+  const ctrl: vscode.TestController = vscode.tests.createTestController(
+    "qsharpTestController",
+    "Q# Tests",
+  );
+  context.subscriptions.push(ctrl);
+
+  const refreshHandler = mkRefreshHandler(ctrl, context);
+  // initially populate tests
+  await refreshHandler(null);
+
+  // when the refresh button is pressed, refresh all tests by passing in a null uri
+  ctrl.refreshHandler = () => refreshHandler(null);
+
+  // when the language service detects an updateDocument, this event fires.
+  // we call the test refresher when that happens
+  updateDocumentEvent(refreshHandler);
+
+  const runHandler = (request: vscode.TestRunRequest) => {
+    if (!request.continuous) {
+      return startTestRun(request);
+    }
+  };
+
+  // runs an individual test run
+  // or test group (a test run where there are child tests)
+  const startTestRun = async (request: vscode.TestRunRequest) => {
+    // use the compiler worker to run the test in the interpreter
+
+    log.trace("Starting test run, request was", JSON.stringify(request));
+    const worker = getCommonCompilerWorker(context);
+
+    const programResult = await getActiveProgram();
+    if (!programResult.success) {
+      throw new Error(programResult.errorMsg);
+    }
+
+    const program = programResult.programConfig;
+
+    for (const testCase of request.include || []) {
+      await runTestCase(ctrl, testCase, request, worker, program);
```

Comment:
I'm not sure I follow the use of the "common worker" here. I thought we discussed the language service could return the list of tests (as this has already parsed/compiled the project on every update and is thus close to zero cost), but the tests should run in an isolated worker for reliability.

Here is seems we're using a common worker to find all the tests (thus on every 'update document' we're compiling the project twice, once for the language service and once for the test discovery - even though most projects will have no tests), and are also running all the tests in that same common worker - not isolated.

Or am I reading this wrong?

Reply:
I see what you're saying now, thanks for typing it out. After our in-person discussion, I had it in my mind that I needed to a) mitigate the worker leak and b) reuse the worker from the language service. I had those two goals in mind when I wrote that code. 

I like the approach of exposing a specific "get test callables" function in the compiler API, because it allows for other scenarios to consume it and write testing infrastructure. Integrating that functionality into the language service to be a language-service-exclusive feature sounds like signing ourselves up for more work in the future.

I think a good approach here would be to modify `compile` such that the returned asset (a `CompileUnit`) contains the list of test callables, and remove `getTestCallables()` elsewhere. Thoughts?